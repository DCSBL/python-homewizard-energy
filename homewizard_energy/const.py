"""Constants for HomeWizard Energy."""

import logging

LOGGER = logging.getLogger(__name__)
