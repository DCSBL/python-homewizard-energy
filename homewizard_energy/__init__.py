"""HomeWizard Energy API library."""

from .errors import DisabledError, InvalidStateError, RequestError, UnsupportedError
from .v1 import HomeWizardEnergyV1
from .v2 import HomeWizardEnergyV2
