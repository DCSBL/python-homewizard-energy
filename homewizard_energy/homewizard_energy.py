"""HomeWizard Energy base class"""

from __future__ import annotations

import asyncio
import json
import ssl
from collections.abc import Callable, Coroutine
from http import HTTPStatus
from typing import Any, TypeVar

import async_timeout
import backoff
from aiohttp.client import (
    ClientError,
    ClientResponseError,
    ClientSession,
    ClientTimeout,
    TCPConnector,
)
from aiohttp.hdrs import METH_DELETE, METH_GET, METH_POST, METH_PUT
from mashumaro.exceptions import InvalidFieldValue, MissingField
from ..models import Device, Measurement, System, SystemUpdate, Token, State

from homewizard_energy.errors import (
    DisabledError,
    RequestError,
    ResponseError,
    UnauthorizedError,
)

class HomeWizardEnergy:
    """Base class for HomeWizard Energy API."""

    _clientsession: ClientSession | None = None
    _request_timeout: int = 10
    _host: str
    
    @property
    def host(self) -> str:
        """Return the hostname of the device.

        Returns:
            host: The used host

        """
        return self._host
    
    async def device(self) -> Device:
        """Get the device information."""
        raise NotImplementedError
    
    async def measurement(self) -> Measurement:
        """Get the current measurement."""
        raise NotImplementedError
    
    async def system(
        self,
        update: SystemUpdate | None = None,
    ) -> System:
        """Get the system information."""
        raise NotImplementedError
        
    async def state(self) -> State:
        """Get the current state."""
        raise NotImplementedError
    
    async def identify(
        self,
    ) -> None:
        """Identify the device."""
        raise NotImplementedError
    
    # async def reboot(
    #     self,
    # ) -> None:
    #     """Reboot the device."""
    #     raise NotImplementedError
