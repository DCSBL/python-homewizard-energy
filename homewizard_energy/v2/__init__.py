"""Representation of a HomeWizard Energy device."""

from __future__ import annotations

import asyncio
import logging
import ssl
from collections.abc import Coroutine
from http import HTTPStatus
from typing import Any, Callable, TypeVar

import aiohttp
import async_timeout
import backoff

from homewizard_energy.errors import (
    DisabledError,
    NotFoundError,
    RequestError,
    ResponseError,
    UnauthorizedError,
)

from .cacert import CACERT
from .models import Device, Measurement, System, SystemUpdate
from .websocket import Websocket

_LOGGER = logging.getLogger(__name__)

T = TypeVar("T")


def authorized_method(
    func: Callable[..., Coroutine[Any, Any, T]],
) -> Callable[..., Coroutine[Any, Any, T]]:
    """Decorator method to check if token is set."""

    async def wrapper(self, *args, **kwargs) -> T:
        # pylint: disable=protected-access
        if self._token is None:
            raise UnauthorizedError("Token missing")

        return await func(self, *args, **kwargs)

    return wrapper


class HomeWizardEnergyV2:
    """Communicate with a HomeWizard Energy device."""

    _clientsession: aiohttp.ClientSession | None = None
    _close_clientsession: bool = False
    _request_timeout: int = 10
    _websocket: Websocket | None = None

    def __init__(
        self,
        host: str,
        identifier: str | None = None,
        token: str | None = None,
        timeout: int = 10,
    ):
        """Create a HomeWizard Energy object.

        Args:
            host: IP or URL for device.
            id: ID for device.
            token: Token for device.
            timeout: Request timeout in seconds.
        """

        self._host = host
        self._identifier = identifier
        self._token = token
        self._request_timeout = timeout

    @property
    def host(self) -> str:
        """Return the hostname of the device.

        Returns:
            host: The used host

        """
        return self._host

    @property
    def websocket(self) -> Websocket:
        """Return the websocket object.

        Create a new websocket object if it does not exist.
        """
        if self._websocket is None:
            self._websocket = Websocket(self)
        return self._websocket

    @authorized_method
    async def device(self) -> Device:
        """Return the device object."""
        _, response = await self._request("/api")
        device = Device.from_dict(response)

        return device

    @authorized_method
    async def measurement(self) -> Measurement:
        """Return the measurement object."""
        _, response = await self._request("/api/measurement")
        measurement = Measurement.from_dict(response)

        return measurement

    @authorized_method
    async def system(
        self,
        update: SystemUpdate | None = None,
    ) -> System:
        """Return the system object."""

        if update is not None:
            data = update.as_dict()
            status, response = await self._request(
                "/api/system", method=aiohttp.hdrs.METH_PUT, data=data
            )

        else:
            status, response = await self._request("/api/system")

        if status != HTTPStatus.OK:
            error = response.get("errodr", response)
            raise RequestError(f"Failed to get system: {error}")

        system = System.from_dict(response)
        return system

    @authorized_method
    async def identify(
        self,
    ) -> bool:
        """Send identify request."""
        await self._request("/api/system/identify", method=aiohttp.hdrs.METH_PUT)
        return True

    async def get_token(
        self,
        name: str,
    ) -> str:
        """Get authorization token from device."""
        status, response = await self._request(
            "/api/user", method=aiohttp.hdrs.METH_POST, data={"name": f"local/{name}"}
        )

        if status == HTTPStatus.FORBIDDEN:
            raise DisabledError("User creation is not enabled on the device")

        if status != HTTPStatus.OK:
            error = response.get("error", response)
            raise RequestError(f"Error occurred while getting token: {error}")

        try:
            token = response["token"]
        except KeyError as ex:
            raise ResponseError("Failed to get token") from ex

        self._token = token
        return token

    @authorized_method
    async def delete_token(
        self,
        name: str | None = None,
    ) -> None:
        """Delete authorization token from device."""
        status, response = await self._request(
            "/api/user",
            method=aiohttp.hdrs.METH_DELETE,
            data={"name": name} if name is not None else None,
        )

        if status != HTTPStatus.NO_CONTENT:
            error = response.get("error", response)
            raise RequestError(f"Error occurred while getting token: {error}")

        # Our token was invalided, resetting it
        if name is None:
            self._token = None

    @property
    def token(self) -> str | None:
        """Return the token of the device.

        Returns:
            token: The used token

        """
        return self._token

    @property
    def request_timeout(self) -> int:
        """Return the request timeout of the device.

        Returns:
            request_timeout: The used request timeout

        """
        return self._request_timeout

    async def get_clientsession(self) -> aiohttp.ClientSession:
        """
        Get a clientsession that is tuned for communication with the HomeWizard Energy Device
        """

        if self._clientsession is not None:
            return self._clientsession

        def _build_ssl_context() -> ssl.SSLContext:
            context = ssl.create_default_context(cadata=CACERT)
            if self._identifier is not None:
                context.hostname_checks_common_name = True
            else:
                _LOGGER.warning("No hostname provided, skipping hostname validation")
                context.check_hostname = False  # Skip hostname validation
                context.verify_mode = ssl.CERT_REQUIRED  # Keep SSL verification active
            return context

        # Creating an SSL context has some blocking IO so need to run it in the executor
        loop = asyncio.get_running_loop()
        context = await loop.run_in_executor(None, _build_ssl_context)

        connector = aiohttp.TCPConnector(
            enable_cleanup_closed=True,
            ssl=context,
            limit_per_host=1,
        )

        self._clientsession = aiohttp.ClientSession(
            connector=connector,
            timeout=aiohttp.ClientTimeout(total=self._request_timeout),
        )

        return self._clientsession

    @backoff.on_exception(backoff.expo, RequestError, max_tries=5, logger=None)
    async def _request(
        self, path: str, method: str = aiohttp.hdrs.METH_GET, data: object = None
    ) -> Any:
        """Make a request to the API."""

        _clientsession = await self.get_clientsession()

        if _clientsession.closed:
            # Avoid runtime errors when connection is closed.
            # This solves an issue when updates were scheduled and clientsession was closed.
            return None

        # Construct request
        url = f"https://{self.host}{path}"
        headers = {
            "Content-Type": "application/json",
        }
        if self._token is not None:
            headers["Authorization"] = f"Bearer {self._token}"

        _LOGGER.debug("%s, %s, %s", method, url, data)

        try:
            async with async_timeout.timeout(self._request_timeout):
                resp = await _clientsession.request(
                    method,
                    url,
                    json=data,
                    headers=headers,
                    server_hostname=self._identifier
                    if self._identifier is not None
                    else None,
                )
                _LOGGER.debug("%s, %s", resp.status, await resp.text("utf-8"))
        except asyncio.TimeoutError as exception:
            raise RequestError(
                f"Timeout occurred while connecting to the HomeWizard Energy device at {self.host}"
            ) from exception
        except (aiohttp.ClientError, aiohttp.ClientResponseError) as exception:
            raise RequestError(
                f"Error occurred while communicating with the HomeWizard Energy device at {self.host}"
            ) from exception

        match resp.status:
            case HTTPStatus.UNAUTHORIZED:
                raise UnauthorizedError("Token rejected")
            case HTTPStatus.METHOD_NOT_ALLOWED:
                raise NotFoundError("Method not allowed")
            case HTTPStatus.NO_CONTENT:
                # No content, just return
                return (HTTPStatus.NO_CONTENT, None)
            case HTTPStatus.OK:
                pass

        content_type = resp.headers.get("Content-Type", "")
        if "application/json" in content_type:
            return (resp.status, await resp.json())

        return (resp.status, await resp.text())

    async def close(self) -> None:
        """Close client session."""
        _LOGGER.debug("Closing clientsession")
        if self._clientsession is not None:
            await self._clientsession.close()
            self._clientsession = None

    async def __aenter__(self) -> HomeWizardEnergyV2:
        """Async enter.

        Returns:
            The HomeWizardEnergyV2 object.
        """
        return self

    async def __aexit__(self, *_exc_info: Any) -> None:
        """Async exit.

        Args:
            _exc_info: Exec type.
        """
        await self.close()
