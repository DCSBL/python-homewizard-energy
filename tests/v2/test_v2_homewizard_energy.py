"""Test for HomeWizard Energy."""

import pytest
from syrupy.assertion import SnapshotAssertion

from homewizard_energy import HomeWizardEnergyV2
from homewizard_energy.errors import DisabledError, UnauthorizedError

from . import load_fixtures

pytestmark = [pytest.mark.asyncio]

### Device tests ###


async def test_class_returns_host():
    """Test class returns host."""

    async with HomeWizardEnergyV2("example.com") as api:
        assert api.host == "example.com"


async def test_device_without_authentication():
    """Test device request is rejected when no authentication is provided."""

    async with HomeWizardEnergyV2("example.com") as api:
        with pytest.raises(UnauthorizedError):
            await api.device()


async def test_device_with_invalid_authentication(aresponses):
    """Test device request is unsuccessful when invalid authentication is provided."""

    aresponses.add(
        "example.com",
        "/api",
        "GET",
        aresponses.Response(
            status=401,
            headers={"Content-Type": "application/json"},
            text='{"error": "user:unauthorized"}',
        ),
    )

    async with HomeWizardEnergyV2("example.com", "token") as api:
        with pytest.raises(UnauthorizedError):
            await api.device()


@pytest.mark.parametrize(
    ("model", "fixtures"),
    [
        ("HWE-P1", ["device"]),
    ],
)
async def test_device_with_valid_authentication(
    model: str, fixtures: str, snapshot: SnapshotAssertion, aresponses
):
    """Test device request is successful when valid authentication is provided."""

    for fixture in fixtures:
        aresponses.add(
            "example.com",
            "/api",
            "GET",
            aresponses.Response(
                text=load_fixtures(f"{model}/{fixture}.json"),
                status=200,
                headers={"Content-Type": "application/json"},
            ),
        )

        async with HomeWizardEnergyV2("example.com", token="token") as api:
            device = await api.device()

            assert device
            assert device.product_type == model

            assert device == snapshot


### Measurement tests ###


async def test_measurement_without_authentication():
    """Test measurement request is rejected when no authentication is provided."""

    async with HomeWizardEnergyV2("example.com") as api:
        with pytest.raises(UnauthorizedError):
            await api.measurement()


async def test_measurement_with_invalid_authentication(aresponses):
    """Test measurement request is unsuccessful when invalid authentication is provided."""

    aresponses.add(
        "example.com",
        "/api/measurement",
        "GET",
        aresponses.Response(
            status=401,
            headers={"Content-Type": "application/json"},
            text='{"error": "user:unauthorized"}',
        ),
    )

    async with HomeWizardEnergyV2("example.com", "token") as api:
        with pytest.raises(UnauthorizedError):
            await api.measurement()


@pytest.mark.parametrize(
    ("model", "fixtures"),
    [
        (
            "HWE-P1",
            [
                "measurement_1_phase_no_gas",
                "measurement_3_phase_with_gas_with_watermeter",
            ],
        ),
    ],
)
async def test_measurement_with_valid_authentication(
    model: str, fixtures: str, snapshot: SnapshotAssertion, aresponses
):
    """Test measurement request is successful when valid authentication is provided."""

    for fixture in fixtures:
        aresponses.add(
            "example.com",
            "/api/measurement",
            "GET",
            aresponses.Response(
                text=load_fixtures(f"{model}/{fixture}.json"),
                status=200,
                headers={"Content-Type": "application/json"},
            ),
        )

        async with HomeWizardEnergyV2("example.com", token="token") as api:
            measurement = await api.measurement()
            assert measurement is not None
            assert measurement == snapshot


### System tests ###


async def test_system_without_authentication():
    """Test system request is rejected when no authentication is provided."""

    async with HomeWizardEnergyV2("example.com") as api:
        with pytest.raises(UnauthorizedError):
            await api.system()


async def test_system_with_invalid_authentication(aresponses):
    """Test system request is unsuccessful when invalid authentication is provided."""

    aresponses.add(
        "example.com",
        "/api/system",
        "GET",
        aresponses.Response(
            status=401,
            headers={"Content-Type": "application/json"},
            text='{"error": "user:unauthorized"}',
        ),
    )

    async with HomeWizardEnergyV2("example.com", "token") as api:
        with pytest.raises(UnauthorizedError):
            await api.system()


@pytest.mark.parametrize(
    ("model", "fixtures"),
    [
        ("HWE-P1", ["system"]),
    ],
)
async def test_system_with_valid_authentication(
    model: str, fixtures: str, snapshot: SnapshotAssertion, aresponses
):
    """Test system request is successful when valid authentication is provided."""

    for fixture in fixtures:
        aresponses.add(
            "example.com",
            "/api/system",
            "GET",
            aresponses.Response(
                text=load_fixtures(f"{model}/{fixture}.json"),
                status=200,
                headers={"Content-Type": "application/json"},
            ),
        )

        async with HomeWizardEnergyV2("example.com", token="token") as api:
            system = await api.system()
            assert system is not None
            assert system == snapshot


### Identify tests ###


async def test_identify_without_authentication():
    """Test identify request is rejected when no authentication is provided."""

    async with HomeWizardEnergyV2("example.com") as api:
        with pytest.raises(UnauthorizedError):
            await api.identify()


async def test_identify_with_invalid_authentication(aresponses):
    """Test identify request is unsuccessful when invalid authentication is provided."""

    aresponses.add(
        "example.com",
        "/api/system/identify",
        "GET",
        aresponses.Response(
            status=401,
            headers={"Content-Type": "application/json"},
            text='{"error": "user:unauthorized"}',
        ),
    )

    async with HomeWizardEnergyV2("example.com", "token") as api:
        with pytest.raises(UnauthorizedError):
            await api.identify()


async def test_identify_with_authentication(aresponses):
    """Test identify request is successful when authentication is provided."""

    aresponses.add(
        "example.com",
        "/api/system/identify",
        "GET",
        aresponses.Response(
            status=204,
            headers={"Content-Type": "application/json"},
        ),
    )

    async with HomeWizardEnergyV2("example.com", "token") as api:
        with pytest.raises(UnauthorizedError):
            data = await api.identify()
            assert data is None


### Identify tests ###


async def test_reboot_without_authentication():
    """Test reboot request is rejected when no authentication is provided."""

    async with HomeWizardEnergyV2("example.com") as api:
        with pytest.raises(UnauthorizedError):
            await api.reboot()


async def test_reboot_with_invalid_authentication(aresponses):
    """Test reboot request is unsuccessful when invalid authentication is provided."""

    aresponses.add(
        "example.com",
        "/api/system/reboot",
        "GET",
        aresponses.Response(
            status=401,
            headers={"Content-Type": "application/json"},
            text='{"error": "user:unauthorized"}',
        ),
    )

    async with HomeWizardEnergyV2("example.com", "token") as api:
        with pytest.raises(UnauthorizedError):
            await api.reboot()


async def test_reboot_with_authentication(aresponses):
    """Test reboot request is successful when authentication is provided."""

    aresponses.add(
        "example.com",
        "/api/system/reboot",
        "GET",
        aresponses.Response(
            status=204,
            headers={"Content-Type": "application/json"},
        ),
    )

    async with HomeWizardEnergyV2("example.com", "token") as api:
        with pytest.raises(UnauthorizedError):
            data = await api.reboot()
            assert data is None


### Get token tests ###


async def test_get_token_without_authentication(aresponses):
    """Test get token request is rejected when no authentication is provided."""

    aresponses.add(
        "example.com",
        "/api/user",
        "POST",
        aresponses.Response(
            status=403,
            headers={"Content-Type": "application/json"},
            text='{"error": "user:creation-not-enabled"}',
        ),
    )

    async with HomeWizardEnergyV2("example.com") as api:
        with pytest.raises(DisabledError):
            await api.get_token("name")


# async def test_request_detects_non_200(aresponses):
#     """Test detects non-ok response."""
#     aresponses.add(
#     """Test detects non-ok response."""
#     aresponses.add(
#         "example.com",
#         "/api",
#         "GET",
#         aresponses.Response(
#             status=500,
#             headers={"Content-Type": "application/json"},
#             text='{"status": "ok"}',
#         ),
#     )
#     async with aiohttp.ClientSession() as session:
#         api = HomeWizardEnergyV1("example.com", clientsession=session)

#         with pytest.raises(RequestError):
#             await api.request("api")

#         await api.close()


# async def test_request_detects_clienterror():
#     """Test other clienterror."""
#     async with aiohttp.ClientSession() as session:
#         api = HomeWizardEnergyV1("example.com", clientsession=session)

#         with (
#             patch.object(session, "request", side_effect=aiohttp.ClientError),
#             pytest.raises(RequestError),
#         ):
#             await api.request("api")

#         await api.close()


# @pytest.mark.parametrize(
#     ("model", "fixtures"),
#     [
#         ("HWE-P1", ["device"]),
#         ("HWE-SKT", ["device"]),
#         ("HWE-WTR", ["device"]),
#         ("HWE-KWH1", ["device"]),
#         ("HWE-KWH3", ["device"]),
#         ("SDM230-wifi", ["device"]),
#         ("SDM630-wifi", ["device"]),
#     ],
# )
# async def test_get_device_object(
#     model: str, fixtures: str, snapshot: SnapshotAssertion, aresponses
# ):
#     """Test device object is fetched and sets detected values."""

#     for fixture in fixtures:
#         aresponses.add(
#             "example.com",
#             "/api",
#             "GET",
#             aresponses.Response(
#                 text=load_fixtures(f"{model}/{fixture}.json"),
#                 status=200,
#                 headers={"Content-Type": "application/json; charset=utf-8"},
#             ),
#         )

#         async with aiohttp.ClientSession() as session:
#             api = HomeWizardEnergyV1("example.com", clientsession=session)
#             device = await api.device()

#             assert device
#             assert device.product_type == model

#             assert device == snapshot

#             await api.close()


# async def test_get_device_object_detects_invalid_api(aresponses):
#     """Test raises error when invalid API is used."""

#     aresponses.add(
#         "example.com",
#         "/api",
#         "GET",
#         aresponses.Response(
#             text=load_fixtures("exceptions/device_invalid_api.json"),
#             status=200,
#             headers={"Content-Type": "application/json; charset=utf-8"},
#         ),
#     )

#     async with aiohttp.ClientSession() as session:
#         api = HomeWizardEnergyV1("example.com", clientsession=session)

#         with pytest.raises(UnsupportedError):
#             await api.device()

#         await api.close()


# @pytest.mark.parametrize(
#     ("model", "fixtures"),
#     [
#         (
#             "HWE-P1",
#             ["data_all_data", "data_minimal", "data_no_gas", "data_single_phase"],
#         ),
#         ("HWE-SKT", ["data"]),
#         ("HWE-WTR", ["data"]),
#         ("HWE-KWH1", ["data"]),
#         ("HWE-KWH3", ["data"]),
#         ("SDM230-wifi", ["data"]),
#         ("SDM630-wifi", ["data"]),
#     ],
# )
# async def test_get_data_object(
#     model: str, fixtures: str, snapshot: SnapshotAssertion, aresponses
# ):
#     """Test fetches data object and device object when unknown."""

#     for fixture in fixtures:
#         aresponses.add(
#             "example.com",
#             "/api",
#             "GET",
#             aresponses.Response(
#                 text=load_fixtures(f"{model}/device.json"),
#                 status=200,
#                 headers={"Content-Type": "application/json; charset=utf-8"},
#             ),
#         )

#         aresponses.add(
#             "example.com",
#             "/api/v1/data",
#             "GET",
#             aresponses.Response(
#                 text=load_fixtures(f"{model}/{fixture}.json"),
#                 status=200,
#                 headers={"Content-Type": "application/json; charset=utf-8"},
#             ),
#         )

#         async with aiohttp.ClientSession() as session:
#             api = HomeWizardEnergyV1("example.com", clientsession=session)

#             data = await api.data()
#             assert data is not None
#             assert data == snapshot

#             await api.close()


# @pytest.mark.parametrize(
#     ("model", "fixtures"),
#     [
#         (
#             "HWE-P1",
#             ["data_all_data", "data_minimal", "data_no_gas", "data_single_phase"],
#         ),
#         ("HWE-SKT", ["data"]),
#         ("HWE-WTR", ["data"]),
#         ("HWE-KWH1", ["data"]),
#         ("HWE-KWH3", ["data"]),
#         ("SDM230-wifi", ["data"]),
#         ("SDM630-wifi", ["data"]),
#     ],
# )
# async def test_get_data_object_with_known_device(
#     model: str, fixtures: str, snapshot: SnapshotAssertion, aresponses
# ):
#     """Test fetches data object."""

#     for fixture in fixtures:
#         aresponses.add(
#             "example.com",
#             "/api/v1/data",
#             "GET",
#             aresponses.Response(
#                 text=load_fixtures(f"{model}/{fixture}.json"),
#                 status=200,
#                 headers={"Content-Type": "application/json; charset=utf-8"},
#             ),
#         )

#         async with aiohttp.ClientSession() as session:
#             api = HomeWizardEnergyV1("example.com", clientsession=session)

#             # pylint: disable=protected-access
#             api._detected_api_version = "v1"

#             data = await api.data()
#             assert data is not None
#             assert data == snapshot

#             await api.close()


# @pytest.mark.parametrize(
#     ("model", "fixtures"),
#     [
#         (
#             "HWE-SKT",
#             ["state_all", "state_power_on", "state_switch_lock", "state_brightness"],
#         ),
#     ],
# )
# async def test_get_state_object(
#     model: str, fixtures: str, snapshot: SnapshotAssertion, aresponses
# ):
#     """Test fetches state object and device object when unknown."""

#     for fixture in fixtures:
#         aresponses.add(
#             "example.com",
#             "/api",
#             "GET",
#             aresponses.Response(
#                 text=load_fixtures(f"{model}/device.json"),
#                 status=200,
#                 headers={"Content-Type": "application/json; charset=utf-8"},
#             ),
#         )

#         aresponses.add(
#             "example.com",
#             "/api/v1/state",
#             "GET",
#             aresponses.Response(
#                 text=load_fixtures(f"{model}/{fixture}.json"),
#                 status=200,
#                 headers={"Content-Type": "application/json; charset=utf-8"},
#             ),
#         )

#         async with aiohttp.ClientSession() as session:
#             api = HomeWizardEnergyV1("example.com", clientsession=session)

#             state = await api.state()
#             assert state is not None
#             assert state == snapshot

#             await api.close()


# @pytest.mark.parametrize(
#     ("model", "fixtures"),
#     [
#         ("HWE-SKT", ["state_all"]),
#     ],
# )
# async def test_state_set(
#     model: str, fixtures: str, snapshot: SnapshotAssertion, aresponses
# ):
#     """Test state set."""

#     for fixture in fixtures:
#         aresponses.add(
#             "example.com",
#             "/api/v1/state",
#             "PUT",
#             aresponses.Response(
#                 text=load_fixtures(f"{model}/{fixture}.json"),
#                 status=200,
#                 headers={"Content-Type": "application/json; charset=utf-8"},
#             ),
#         )

#         async with aiohttp.ClientSession() as session:
#             api = HomeWizardEnergyV1("example.com", clientsession=session)

#             state = await api.state_set(
#                 power_on=False, switch_lock=False, brightness=255
#             )
#             assert state

#             assert state == snapshot

#             await api.close()


# async def test_state_set_detects_no_statechange():
#     """Test state set does not send request when nothing is changed."""

#     async with aiohttp.ClientSession() as session:
#         api = HomeWizardEnergyV1("example.com", clientsession=session)

#         state = await api.state_set()
#         assert not state


# @pytest.mark.parametrize(
#     "model",
#     [
#         "HWE-P1",
#         "HWE-SKT",
#         "HWE-WTR",
#     ],
# )
# async def test_identify(model: str, snapshot: SnapshotAssertion, aresponses):
#     """Test identify call."""

#     aresponses.add(
#         "example.com",
#         "/api/v1/identify",
#         "PUT",
#         aresponses.Response(
#             text=load_fixtures(f"{model}/identify.json"),
#             status=200,
#             headers={"Content-Type": "application/json; charset=utf-8"},
#         ),
#     )

#     async with aiohttp.ClientSession() as session:
#         api = HomeWizardEnergyV1("example.com", clientsession=session)

#         state = await api.identify()
#         assert state
#         assert state == snapshot

#         await api.close()


# @pytest.mark.parametrize(
#     "model",
#     [
#         "HWE-KWH1",
#         "HWE-KWH3",
#         "SDM230-wifi",
#         "SDM630-wifi",
#     ],
# )
# async def test_identify_not_supported(model: str, aresponses):
#     """Test identify call when not supported."""

#     aresponses.add(
#         "example.com",
#         "/api",
#         "GET",
#         aresponses.Response(
#             text=load_fixtures(f"{model}/device.json"),
#             status=200,
#             headers={"Content-Type": "application/json; charset=utf-8"},
#         ),
#     )

#     aresponses.add(
#         "example.com",
#         "/api/v1/identify",
#         "PUT",
#         aresponses.Response(
#             text="404 Not Found",
#             status=404,
#             headers={"Content-Type": "application/txt; charset=utf-8"},
#         ),
#     )

#     async with aiohttp.ClientSession() as session:
#         api = HomeWizardEnergyV1("example.com", clientsession=session)

#         with pytest.raises(UnsupportedError):
#             await api.identify()

#         await api.close()


# async def test_get_system_object(aresponses):
#     """Test fetches system object and device object when unknown."""

#     aresponses.add(
#         "example.com",
#         "/api",
#         "GET",
#         aresponses.Response(
#             text=load_fixtures("HWE-SKT/device.json"),
#             status=200,
#             headers={"Content-Type": "application/json; charset=utf-8"},
#         ),
#     )

#     aresponses.add(
#         "example.com",
#         "/api/v1/system",
#         "GET",
#         aresponses.Response(
#             text=json.dumps({"cloud_enabled": True}),
#             status=200,
#             headers={"Content-Type": "application/json; charset=utf-8"},
#         ),
#     )

#     aresponses.add(
#         "example.com",
#         "/api/v1/system",
#         "GET",
#         aresponses.Response(
#             text=json.dumps({"cloud_enabled": False}),
#             status=200,
#             headers={"Content-Type": "application/json; charset=utf-8"},
#         ),
#     )

#     async with aiohttp.ClientSession() as session:
#         api = HomeWizardEnergyV1("example.com", clientsession=session)

#         system = await api.system()
#         assert system
#         assert system.cloud_enabled

#         system = await api.system()
#         assert system
#         assert not system.cloud_enabled

#         await api.close()


# @pytest.mark.parametrize(
#     "model",
#     [
#         "HWE-P1",
#         "HWE-SKT",
#         "HWE-WTR",
#         "HWE-KWH1",
#         "HWE-KWH3",
#         "SDM230-wifi",
#         "SDM630-wifi",
#     ],
# )
# async def test_system_set(model: str, snapshot: SnapshotAssertion, aresponses):
#     """Test system set."""

#     aresponses.add(
#         "example.com",
#         "/api/v1/system",
#         "PUT",
#         aresponses.Response(
#             text=load_fixtures(f"{model}/system.json"),
#             status=200,
#             headers={"Content-Type": "application/json; charset=utf-8"},
#         ),
#     )

#     async with aiohttp.ClientSession() as session:
#         api = HomeWizardEnergyV1("example.com", clientsession=session)

#         system = await api.system_set(cloud_enabled=False)
#         assert system
#         assert system == snapshot

#         await api.close()


# async def test_system_set_missing_arguments():
#     """Test system set when no arguments are given."""

#     async with aiohttp.ClientSession() as session:
#         api = HomeWizardEnergyV1("example.com", clientsession=session)
#         assert await api.system_set() is False


# # pylint: disable=protected-access
# async def test_request_timeout():
#     """Test request raises timeout when request takes too long."""

#     api = HomeWizardEnergyV1("example.com")
#     api._session = AsyncMock()
#     api._session.request = AsyncMock(side_effect=asyncio.TimeoutError())

#     with pytest.raises(RequestError):
#         await api.request("api/v1/data")

#     assert api._session.request.call_count == 5


# async def test_close_when_out_of_scope():
#     """Test close called when object goes out of scope."""
#     api = HomeWizardEnergyV1("example.com")
#     api.close = AsyncMock()

#     async with api as hwe:
#         assert hwe == api

#     assert api.close.call_count == 1
