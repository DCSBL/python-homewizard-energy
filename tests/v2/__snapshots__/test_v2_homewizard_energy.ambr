# serializer version: 1
# name: test_device_with_valid_authentication[HWE-P1-fixtures0]
  Device(product=Product(model='HWE-P1', name='Wi-Fi P1 Meter', url='https://www.homewizard.com/p1-meter/', description='The HomeWizard P1 Meter gives you detailed insight in your electricity-, gas consumption and solar surplus.'), product_name='P1 Meter', product_type='HWE-P1', serial='5c2fafaabbcc', api_version='2.0.0', firmware_version='6.00')
# ---
# name: test_measurement_with_valid_authentication[HWE-P1-fixtures0]
  Measurement(protocol_version=None, meter_model='ISKRA  2M550T-101', unique_id='Never Gonna Give You Up', tariff=None, energy_import_kwh=None, energy_import_t1_kwh=10830.511, energy_import_t2_kwh=2948.827, energy_import_t3_kwh=None, energy_import_t4_kwh=None, energy_export_kwh=None, energy_export_t1_kwh=1285.951, energy_export_t2_kwh=2876.51, energy_export_t3_kwh=None, energy_export_t4_kwh=None, power_w=-678, power_l1_w=-676, power_l2_w=None, power_l3_w=None, voltage_l1_v=None, voltage_l2_v=None, voltage_l3_v=None, current_a=None, current_l1_a=None, current_l2_a=None, current_l3_a=None, frequency_hz=None, voltage_sag_l1_count=None, voltage_sag_l2_count=None, voltage_sag_l3_count=None, voltage_swell_l1_count=None, voltage_swell_l2_count=None, voltage_swell_l3_count=None, any_power_fail_count=None, long_power_fail_count=None, average_power_15m_w=None, monthly_power_peak_w=None, monthly_power_peak_timestamp=None, external_devices=None)
# ---
# name: test_measurement_with_valid_authentication[HWE-P1-fixtures0].1
  Measurement(protocol_version=None, meter_model='ISKRA  2M550T-101', unique_id='Never Gonna Let You Down', tariff=None, energy_import_kwh=13779.338, energy_import_t1_kwh=10830.511, energy_import_t2_kwh=2948.827, energy_import_t3_kwh=None, energy_import_t4_kwh=None, energy_export_kwh=0, energy_export_t1_kwh=0, energy_export_t2_kwh=0, energy_export_t3_kwh=None, energy_export_t4_kwh=None, power_w=-543, power_l1_w=-676, power_l2_w=133, power_l3_w=0, voltage_l1_v=None, voltage_l2_v=None, voltage_l3_v=None, current_a=6, current_l1_a=-4, current_l2_a=2, current_l3_a=0, frequency_hz=None, voltage_sag_l1_count=1, voltage_sag_l2_count=1, voltage_sag_l3_count=0, voltage_swell_l1_count=0, voltage_swell_l2_count=0, voltage_swell_l3_count=0, any_power_fail_count=4, long_power_fail_count=5, average_power_15m_w=123.0, monthly_power_peak_w=1111.0, monthly_power_peak_timestamp=datetime.datetime(2024, 6, 4, 10, 11, 22), external_devices={'gas_meter_Never gonna run around': ExternalDevice(unique_id='Never gonna run around', meter_type=<DeviceType.GAS_METER: 3>, value=2569.646, unit='m3', timestamp=datetime.datetime(2024, 6, 28, 14, 0)), 'water_meter_and desert you': ExternalDevice(unique_id='and desert you', meter_type=<DeviceType.WATER_METER: 7>, value=123.456, unit='m3', timestamp=datetime.datetime(2024, 6, 28, 14, 5))})
# ---
# name: test_system_with_valid_authentication[HWE-P1-fixtures0]
  System(wifi_ssid='My Wi-Fi', wifi_rssi_db=-77, cloud_enabled=False, uptime_s=356, status_led_brightness_pct=100, api_v1_enabled=True)
# ---
